src/extension/panel.ts

```diff
@@ -0,0 +1,162 @@
+    constructor(
+        readonly context: Pick<ExtensionContext, 'extensionPath' | 'subscriptions'>,
+        readonly basing: Baser,
+        readonly store: Pick<Store, 'logs' | 'results'>) {
+        observe(store.logs, change => {
+            const {type, removed, added} = change as unknown as IArraySplice<Log>;
+            if (type !== 'splice') throw new Error('Only splice allowed on store.logs.');
+            this.spliceLogs(removed, added);
+        });
+        autorun(() => {
+            const count = store.results.length;
+            if (!this.panel) return;
+            this.panel.title = `${count} ${this.title}${count === 1 ? '' : 's'}`;
+        });
+    }
+
+    public async show() {
+        if (this.panel) {
+            if (!this.panel.active) this.panel.reveal(undefined, true);
+            return;
+        }
+
+        const {context, basing, store} = this;
+        const {webview} = this.panel = window.createWebviewPanel(
+            'Index', `${this.title}s`, { preserveFocus: true, viewColumn: ViewColumn.Two }, // ViewColumn.Besides steals focus regardless of preserveFocus.
+            {
+                enableScripts: true,
+                localResourceRoots: [Uri.file('/'), ...'abcdefghijklmnopqrstuvwxyz'.split('').map(c => Uri.file(`${c}:`))],
+                retainContextWhenHidden: true,
+            }
+        );
+        this.panel.onDidDispose(() => this.panel = null);
+
+        const src = Uri.file(`${context.extensionPath}/out/panel.js`);
+        const defaultState = {
+            version: 0,
+            filtersRow,
+            filtersColumn,
+        };
+        const state = Store.globalState.get('view', defaultState);
+        webview.html = `<!DOCTYPE html>
+            <html lang="en">
+            <head>
+                <meta charset="UTF-8">
+                <meta http-equiv="Content-Security-Policy" content="
+                    default-src 'none';
+                    connect-src vscode-resource:;
+                    font-src    data:;
+                    script-src  vscode-resource: 'unsafe-inline';
+                    style-src   vscode-resource: 'unsafe-inline';
+                    ">
+                <style>
+                    code { font-family: ${workspace.getConfiguration('editor').get('fontFamily')} }
+                </style>
+            </head>
+            <body>
+                <div id="root"></div>
+                <script src="${webview.asWebviewUri(src).toString()}"></script>
+                <script>
+                    vscode = acquireVsCodeApi()
+                    ReactDOM.render(
+                        React.createElement(Index, { store: new Store(${JSON.stringify(state)}) }),
+                        document.getElementById('root'),
+                    )
+                </script>
+            </body>
+            </html>`;
+
+        webview.onDidReceiveMessage(async message => {
+            if (!message) return;
+            switch (message.command) {
```

Comment:
>command [](start = 28, length = 7)

Let's make sure the "command" here is casted to a strong type so in the future we don't accidentally mis-type the command name.

Reply:
Added. Also symmetrically added it for indexStore.ts. Note, unfortunately the VS Code API to send the message isn't generic, and I haven't gone through the trouble to wrap it.

---
In reply to: [441686829](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441686829) [](ancestors = 441686829)